### ISSUE TITLE
Possible to use pointer types via iterators without copy construction?

### ISSUE THREAD

**Issue by eclarkso** [fnc12/sqlite_orm#776]
This is a sort of followup to a similar the question/issue in https://github.com/fnc12/sqlite_orm/issues/500...

Similarly, I have a data table with many objects in it, *and* those objects have some pretty heavy fields.
```c++
struct DataProxy
{
	// we could define this if necessary, but doesn't help issue below
//	explicit DataProxy(DataProxy&& other) = default; 
	explicit DataProxy() = default;
	explicit DataProxy(const Node * node, const std::string & _file);

	uint id;
	std::string file;
	std::vector<char> data; // could be several MB BLOB
// FK - following example in https://github.com/fnc12/sqlite_orm/blob/master/examples/foreign_key.cpp
	std::unique_ptr<uint> nodeID; 
}
```
As with the linked issue, I want to make available to a caller an iterator (so that I don't have to store all the `DataProxy` objects in memory).  But I also want to avoid (expensive) copies of each `DataProxy` - and regardless, there is no `DataProxy` copy constructor because of the `nodeID` unique pointer (foreign key field).

The problem is that the iterator seems to require an object copy constructor:
```c++
// in a DB manager singleton
auto getData(std::set<uint> ids)
{
	using namespace sqlite_orm;
	return storage->iterate<DataProxy>(join<NodeProxy>(on(c(&NodeProxy::id) == &DataProxy::nodeID)),
	                                   where(in(c(& NodeProxy::id),ids)));
}

// elsewhere
auto dataIterator = DatabaseManager::instance()->getData(ids);

// compiler error: call to implictly-deleted copy constructor of DataProxy		
for(auto dataProxy : dataIterator) {
	// hopefully use proxy object *without* copies 
}
```

Obviously as a workaround I can just implement a copy constructor - but perhaps am I missing something obvious (maybe general c++ knowledge)?  Are there other `sqlite_orm` iterator methods I've missed?  Can I structure my objects or code differently? If so, it's not obvious to me how...

Thanks for the ORM library, it has been great to use.

**Comment by fnc12** [comment]
Just replace 

'for(auto dataProxy : dataIterator) {'

with

'for(auto &dataProxy : dataIterator) {'

**Comment by eclarkso** [comment]
I was just coming here to say I missed the reference in the example (🤦).  Sorry about that - I'm still having compile issues but I suspect they're stuff I just need to work through.

**Comment by eclarkso** [comment]
Does `iterate` support `join`, and if so what am I doing wrong?  A simple `storage->iterate<DataProxy>()` compiles, but if I try a `join` (with `where`)
```c++
storage->iterate<DataProxy>(join<NodeProxy>(on(c(&NodeProxy::id) == &DataProxy::nodeID)),
	                                   where(in(c(&NodeProxy::id),ids)));
```
I get confusing-to-me c++/sqlite_orm template specialization compile (clang) errors:
```
...sqlite_orm/include/sqlite_orm/sqlite_orm.h:9581:39: error: implicit instantiation of undefined template 'sqlite_orm::internal::statement_serializator<sqlite_orm::internal::expression_t<unsigned int NodeProxy::*>>'
            statement_serializator<T> serializator;
                                      ^
...sqlite_orm/include/sqlite_orm/sqlite_orm.h:10034:35: note: in instantiation of function template specialization 'sqlite_orm::internal::serialize<........giant template type listing............>>>>>>' requested here
                auto leftString = serialize(c.l, context);
                                  ^
```
Also perhaps relevant, since they appear in the giant type listing above - there are two other DB structures/tables (`TestcaseProxy` and `TestMatrixProxy`).  I don't think anything is special about them - they are similar to `NodeProxy` in that they have an `id` PK and have a FK in the table "below" it, so the 1:N hierarchy is TestMatrix -> Testcase -> Node -> Data.  But neither of them affect what I'm trying to do (yet - but I assume if I can get one join to work I can get >1).

**Comment by fnc12** [comment]
storage->iterate<DataProxy>(join<NodeProxy>(on(c(&NodeProxy::id) == &DataProxy::nodeID)),
	                                   where(in(c(&NodeProxy::id),ids)));

How would you write this query using sql?

**Comment by eclarkso** [comment]
```sql
SELECT * FROM data
JOIN node ON node.id = data.nodeID
WHERE node.id IN (...)
```
I realize in this particular case, I could omit the join and just do a simple where on the `nodeID` field (which you can do per the `iterate` example).  

However, what I'm *really* after is a query like
```sql
SELECT * FROM data
JOIN node ON node.id = data.nodeID
JOIN testcase ON testcase.id = node.testcaseID
JOIN testmatrix ON testmatris.id = testcase.testmatrixID
WHERE testmatrix.id IN (...)
```
... where I need the JOINs.

**Comment by fnc12** [comment]
AFAIK `JOIN` must be placed after `WHERE`

**Comment by fnc12** [comment]
also replace `in(c(&NodeProxy::id),ids)` with `in(&NodeProxy::id, ids)`

**Comment by fnc12** [comment]
don't use `c` function if you are not sure that you need to use it

**Comment by eclarkso** [comment]
Ok, tried
```c++
storage->iterate<DataProxy>(where(in(&NodeProxy::id,ids)), // where first, no `c()` usage
                            join<NodeProxy>(on(c(&NodeProxy::id) == &DataProxy::nodeID)));
```
but get a similar gigantic undefined template error.  The first bit (line 9581) is the same, but the 2nd is from line 10038 instead of 10034.  I'm guessing that's from getting rid of the extra `c()` usage though:
```
sqlite_orm/include/sqlite_orm/sqlite_orm.h:9581:39: error: implicit instantiation of undefined template 'sqlite_orm::internal::statement_serializator<std::__1::set<unsigned int>>'
            statement_serializator<T> serializator;
                                      ^
/Users/ec113/Code/git/tmt2/src/external/sqlite_orm/include/sqlite_orm/sqlite_orm.h:10038:23: note: in instantiation of function template specialization 'sqlite_orm::internal::serialize<std::__1::set<unsigned int>, sqlite_orm::internal::serializator_context<sqlite_orm::internal::storage_impl<sqlite_orm::internal::index_t<sqlite_orm::internal::indexed_column_t<std::__1::basic_string<char> DataProxy::*>>, sqlite_orm::internal::table_t<TestMatrixProxy, sqlite_orm::internal::column_t<TestMatrixProxy, unsigned int, const unsigned int &(TestMatrixProxy::*)() const, void (TestMatrixProxy::*)(unsigned int), sqlite_orm::constraints::autoincrement_t, sqlite_orm::constraints::primary_key_t<>>, sqlite_orm::internal::column_t<TestMatrixProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestMatrixProxy::*)() const, void (TestMatrixProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<TestMatrixProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestMatrixProxy::*)() const, void (TestMatrixProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<TestMatrixProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestMatrixProxy::*)() const, void (TestMatrixProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<TestMatrixProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestMatrixProxy::*)() const, void (TestMatrixProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<TestMatrixProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestMatrixProxy::*)() const, void (TestMatrixProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<TestMatrixProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestMatrixProxy::*)() const, void (TestMatrixProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<TestMatrixProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestMatrixProxy::*)() const, void (TestMatrixProxy::*)(std::__1::basic_string<char>)>>, sqlite_orm::internal::table_t<TestcaseProxy, sqlite_orm::internal::column_t<TestcaseProxy, unsigned int, const unsigned int &(TestcaseProxy::*)() const, void (TestcaseProxy::*)(unsigned int), sqlite_orm::constraints::autoincrement_t, sqlite_orm::constraints::primary_key_t<>>, sqlite_orm::internal::column_t<TestcaseProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestcaseProxy::*)() const, void (TestcaseProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<TestcaseProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(TestcaseProxy::*)() const, void (TestcaseProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<TestcaseProxy, std::__1::unique_ptr<unsigned int>, const std::__1::unique_ptr<unsigned int> &(TestcaseProxy::*)() const, void (TestcaseProxy::*)(std::__1::unique_ptr<unsigned int>)>, sqlite_orm::constraints::foreign_key_t<std::__1::tuple<std::__1::unique_ptr<unsigned int> TestcaseProxy::*>, std::__1::tuple<unsigned int TestMatrixProxy::*>>>, sqlite_orm::internal::table_t<NodeProxy, sqlite_orm::internal::column_t<NodeProxy, unsigned int, const unsigned int &(NodeProxy::*)() const, void (NodeProxy::*)(unsigned int), sqlite_orm::constraints::autoincrement_t, sqlite_orm::constraints::primary_key_t<>>, sqlite_orm::internal::column_t<NodeProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<NodeProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<NodeProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<NodeProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<NodeProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<NodeProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<NodeProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<NodeProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<NodeProxy, int, const int &(NodeProxy::*)() const, void (NodeProxy::*)(int)>, sqlite_orm::internal::column_t<NodeProxy, bool, const bool &(NodeProxy::*)() const, void (NodeProxy::*)(bool)>, sqlite_orm::internal::column_t<NodeProxy, std::__1::unique_ptr<unsigned int>, const std::__1::unique_ptr<unsigned int> &(NodeProxy::*)() const, void (NodeProxy::*)(std::__1::unique_ptr<unsigned int>)>, sqlite_orm::constraints::foreign_key_t<std::__1::tuple<std::__1::unique_ptr<unsigned int> NodeProxy::*>, std::__1::tuple<unsigned int TestcaseProxy::*>>>, sqlite_orm::internal::table_t<DataProxy, sqlite_orm::internal::column_t<DataProxy, std::__1::basic_string<char>, const std::__1::basic_string<char> &(DataProxy::*)() const, void (DataProxy::*)(std::__1::basic_string<char>)>, sqlite_orm::internal::column_t<DataProxy, std::__1::vector<char>, const std::__1::vector<char> &(DataProxy::*)() const, void (DataProxy::*)(std::__1::vector<char>)>, sqlite_orm::internal::column_t<DataProxy, std::__1::unique_ptr<unsigned int>, const std::__1::unique_ptr<unsigned int> &(DataProxy::*)() const, void (DataProxy::*)(std::__1::unique_ptr<unsigned int>)>, sqlite_orm::constraints::foreign_key_t<std::__1::tuple<std::__1::unique_ptr<unsigned int> DataProxy::*>, std::__1::tuple<unsigned int NodeProxy::*>>>>>>' requested here
                ss << serialize(c.arg, newContext);
```
I'm including the entire gigantic template to show the presence of `TestcaseProxy` and `TestMatrixProxy` in the template error, which doesn't make much sense to me since they're irrelevant to the query.  I tried the iterate call also joining those two tables  to see if that had an effect, but nope.

**Comment by fnc12** [comment]
looks like `ids` has type `std::set` which cannot be serialized. Either we need to add serialization for it into `sqlite_orm` or you need to change it to `std::vector`

**Comment by eclarkso** [comment]
Aaaah, I feel so dumb.  `std::set` serialization would be useful for sure, but that's minor.  To summarize what I learned for anyone finding this in search:

- `iterator` returns a list of references
- `where` must come *before* `join`
- only use `c()` method when required (e.g., `c(&C::foo) == value`)
- only use `std::vector` containers

Thank you so much for your help - and sorry for creating an issue too, after the fact I saw it should be a "Discussion" topic.
